subprojects/language-groovy/src/main/java/org/gradle/api/tasks/compile/GroovyCompile.java

```diff
@@ -62,23 +100,119 @@
+    private final ConfigurableFileCollection astTransformClasspath;
+    private final File sourceClassesMappingFile;
+    private final FileCollection stableSources = getProject().files(new Callable<FileTree>() {
+        @Override
+        public FileTree call() {
+            return getSource();
+        }
+    });
 
     public GroovyCompile() {
-        CompileOptions compileOptions = getServices().get(ObjectFactory.class).newInstance(CompileOptions.class);
+        ObjectFactory objectFactory = getServices().get(ObjectFactory.class);
+        CompileOptions compileOptions = objectFactory.newInstance(CompileOptions.class);
+        compileOptions.setIncremental(false);
         this.compileOptions = compileOptions;
+        this.astTransformClasspath = objectFactory.fileCollection();
         CompilerForkUtils.doNotCacheIfForkingViaExecutable(compileOptions, getOutputs());
+        boolean enableCompileAvoidance = Boolean.getBoolean("org.gradle.unsafe.groovy.compile.avoidance");
+        if (!enableCompileAvoidance) {
+            astTransformClasspath.from(new Callable<FileCollection>() {
+                @Override
+                public FileCollection call() {
+                    return getClasspath();
+                }
+            });
+        }
+        this.sourceClassesMappingFile = new File(getTemporaryDir(), "source-classes-mapping.txt");
     }
 
     @Override
-    @TaskAction
     protected void compile() {
         checkGroovyClasspathIsNonEmpty();
         DefaultGroovyJavaJointCompileSpec spec = createSpec();
-        WorkResult result = getCompiler(spec).execute(spec);
+        WorkResult result = getCompiler(null, spec, false, null).execute(spec);
         setDidWork(result.getDidWork());
+        updateMappingsFile(spec.getCompilationMappingFile(), ImmutableMultimap.<File, String>of());
+    }
+
+    @TaskAction
+    protected void compile(InputChanges inputChanges) {
+        System.out.println("Incremental compilation: " + inputChanges.isIncremental());
+        if (!getOptions().isIncremental()) {
+            compile();
+            return;
+        }
+
+        if (!inputChanges.isIncremental() || !sourceClassesMappingFile.isFile()) {
+            compile();
+        } else {
+            Multimap<File, String> sourceClassesMapping = readSourceClassesMapping(sourceClassesMappingFile);
+            DefaultGroovyJavaJointCompileSpec spec = createSpec();
+            WorkResult result = getCompiler(inputChanges, spec, true, sourceClassesMapping).execute(spec);
+            setDidWork(result.getDidWork());
+            updateMappingsFile(spec.getCompilationMappingFile(), sourceClassesMapping);
+        }
+    }
+
+    private void updateMappingsFile(File compilationMappingFile, Multimap<File, String> oldMapping) {
+        Multimap<File, String> mapping = MultimapBuilder.hashKeys().arrayListValues().build();
+        mapping.putAll(oldMapping);
```

Comment:
This is only adding stuffs. What if some files are deleted in oldMapping? It will never be cleaned up right?